benchmark/src/steps.ts

```diff
@@ -0,0 +1,265 @@
+    /**
+     * The end result of the state export migration
+     */
+    public readonly migrate?: BenchmarkValue<StateExport>;
+    /**
+     * The end result of the exercise validation
+     * The value is an array of the errors
+     */
+    public readonly validateExercise?: BenchmarkValue<
+        ReturnType<typeof validateExerciseExport>
+    >;
+    /**
+     * The frozen state of the exercise
+     */
+    public readonly freezeState?: {
+        initialState: ExerciseState;
+        actionHistory: readonly ExerciseAction[];
+    };
+    // There are three different ways to create the end state of the exercise
+    public readonly newImmerDraft?: BenchmarkValue<ExerciseState>;
+    public readonly sameImmerDraft?: BenchmarkValue<ExerciseState>;
+    public readonly noImmerDraft?: BenchmarkValue<ExerciseState>;
+    /**
+     * Whether the end states of all three methods to create them are equal
+     */
+    public readonly endStatesAreEqual?: boolean;
+    /**
+     * The total benchmarked time it took to execute one action of each type in ms
+     * (sum of all benchmarked times)
+     * Sorted by time (descending)
+     */
+    public readonly benchmarkActions?: {
+        [key in ExerciseAction['type']]?: number;
+    };
+    /**
+     * A string with the most expensive action in the exercise (by total summed-up time) and a representation of the respective time
+     */
+    public readonly mostExpensiveAction?: string;
+    /**
+     * The number of actions of each type
+     * Sorted by amount (descending)
+     */
+    public readonly numberOfActionsPerType?: {
+        [key in ExerciseAction['type']]?: number;
+    };
+    /**
+     * The number of actions in the exercise
+     */
+    // The key is used as the column-name in the table, therefore the weird name
+    public readonly '#actions'?: number;
+
+    constructor(public readonly data: StateExport) {}
+}
+
+/**
+ * The steps are executed for each state export.
+ * The steps are executed in the order they are defined.
+ * The result of each step is stored in `stepState[stateName]`.
+ * Each step can only access the results of the previous steps.
+ *`stepState.data` is always available.
+ *
+ * Steps can print to the console and throw errors to stop the execution of the specific exercise-state.
+ * In addition, at the end of the benchmark, there is a summary of the results in the form of a table.
+ * A lot of this is specified in the individual step classes.
+ */
+export const steps: Step<StepState>[] = [
+    new BenchmarkStep(
+        'migrate',
+        ({ data }) => migrateStateExport(data) as StateExport
+    ),
+    new BenchmarkStep('validateExercise', ({ migrate: migratedValues }) =>
+        validateExerciseExport(migratedValues!.value)
+    ),
+    new CalculationStep(
+        'freezeState',
+        ({ migrate: migratedValues }) => {
+            if (!migratedValues?.value.history) {
+                throw new Error('State export is missing history');
```

Comment:
While I understand that for the next steps it's required to have a history, I think it would be a good idea to allow exports without it, especially to benchmark imports (incl. migrations) of exports without history. So I'd like to have a way for this to fail gracefully.

Reply:
While it still doesn't fail gracefully, I changed it to still print out all the results collected during the successful steps in the end result table.
This should be a good enough compromise.